tests/vmi_networking_test.go

```diff
@@ -863,3 +860,51 @@ func NewRandomVMIWithInvalidNetworkInterface() *v1.VirtualMachineInstance {
 	vmi.Spec.Domain.Devices.Interfaces[0].Model = "gibberish"
 	return vmi
 }
+
+func createExpectTraceroute6(address string) []expect.Batcher {
+	return []expect.Batcher{
+		&expect.BSnd{S: "\n"},
+		&expect.BExp{R: "\\$ "},
+		&expect.BSnd{S: "traceroute -6 " + address + " -w1 > tr\n"},
+		&expect.BExp{R: "\\$ "},
+		&expect.BSnd{S: "cat tr | grep -q \"*\\|!\"\n"},
+		&expect.BExp{R: "\\$ "},
+		&expect.BSnd{S: "echo $?\n"},
+		&expect.BExp{R: "1"},
+	}
+}
+
+func createExpectStartTcpServer(port string) []expect.Batcher {
+	return []expect.Batcher{
+		&expect.BSnd{S: "\n"},
+		&expect.BExp{R: "\\$ "},
+		&expect.BSnd{S: "screen -d -m sudo nc -klp " + port + " -e echo -e 'Hello World!'\n"},
+		&expect.BExp{R: "\\$ "},
+		&expect.BSnd{S: "echo $?\n"},
+		&expect.BExp{R: "0"},
+	}
+}
+
+func createExpectConnectToServer(serverIP, tcpPort string, expectSuccess bool) []expect.Batcher {
+	expectResult := "1"
+	if expectSuccess {
+		expectResult = "0"
+	}
+	return []expect.Batcher{
+		&expect.BSnd{S: "\n"},
+		&expect.BExp{R: "\\$ "},
+		&expect.BSnd{S: fmt.Sprintf("echo test | nc %s %s -i 1 -w 1 1> /dev/null\n", serverIP, tcpPort)},
+		&expect.BExp{R: "\\$ "},
+		&expect.BSnd{S: "echo $?\n"},
+		&expect.BExp{R: expectResult},
+	}
+}
+
+// gatewayIpFromCIDR returns the first address of a network.
+func gatewayIPFromCIDR(cidr string) string {
+	ip, ipnet, _ := net.ParseCIDR(cidr)
+	ip = ip.Mask(ipnet.Mask)
+	oct := len(ip) - 1
+	ip[oct]++
```

Comment:
Looks nicer, thanks.
But if the last byte of the ip is 'ff' it won't work (please refer to [inc](https://github.com/kubevirt/kubevirt/blob/51ce9d8e6def33c0c260f0cdb994f08ae5b9938e/pkg/virt-launcher/virtwrap/network/common.go#L238)).

Anyway, I would pass to the test the desired ip address of the gateway explicitly. The goal of the test is to verify the flow works as we expect. By duplicating the code, I think the goal is somehow missed.

Reply:
We do not need to handle the edge cases, it is a helper for the test. If a bad input is given, the test will constantly fail.
BTW, because we increment only once, if someone will pass `fc10::ff/120` or something like that, the result will still be ending with `01` because of the mask.

The reason we need this is to fit both default and custom values and if the default change (like it already did with IPv6) I do not want to touch here. This is also minimal to worth it IMO.

Reply:
Well yes, here you increment only once, so it should be ok.

The default CIDR is something that is not expected to be changed. And if it is, it makes sense the test should be updated.
And the statement 'If a bad input is given, the test will constantly fail' is not valid. Since the test calculates the gateway, if the calculation has the same bug the code has, the test won't fail.

But since unit tests should catch wrong calculation of the gateway in the code and the logic is pretty simple. I won't block you on it.

I saw you were asked to include the ipv6 cidr test in this PR and not a separate one. So I'm waiting for it before approving the PR.

Reply:
> I saw you were asked to include the ipv6 cidr test in this PR and not a separate one. So I'm waiting for it before approving the PR.

Where I was asked to do that? And I cannot do it here because this is IPv4 only PR, the IPv6 ones will come with `https://github.com/kubevirt/kubevirt/pull/3383` or a following one that adds tests.

Reply:
You're right, my mistake. I was confused with #3383 that doesn't have the tests yet.

Reply:
I can add there tests when @maiqueb PR gets in (https://github.com/kubevirt/kubevirt/pull/3454).

Reply:
According to Miguel's comment on the PR. The PR is relevant for fedora VMs.
This test is using cirros image.

Reply:
I will change it :)